### PR TITLE
mvn compl

### PR DESCRIPTION
**Pull request by qreqit**

### CODE REVIEW

File: src/main/java/mate/academy/springbootwebgreqit/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    @Bean
+    public PasswordEncoder getPasswordEncoder() {
+        return new BCryptPasswordEncoder();
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        http
+            .authorizeHttpRequests(authorize -> authorize
+                .requestMatchers("/auth/registration", "/swagger-ui/**", "/v3/api-docs/**").permitAll()
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/config/SecurityConfig.java:23]:
```suggestion
                .requestMatchers(
                                        antMatcher("/auth/**"),
                                        antMatcher("/swagger-ui/**"),
                                        antMatcher("/v3/api-docs/**"))
                .permitAll()
                anyRequest()
               .authenticated()
```

File: src/main/java/mate/academy/springbootwebgreqit/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+public class SecurityConfig {
+    @Bean
+    public PasswordEncoder getPasswordEncoder() {
+        return new BCryptPasswordEncoder();
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        http
+            .authorizeHttpRequests(authorize -> authorize
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/config/SecurityConfig.java:22]:
```suggestion
            .cors(AbstractHttpConfigurer::disable)
            .csrf(AbstractHttpConfigurer::disable)
            .authorizeHttpRequests(authorize -> authorize
```

File: src/main/java/mate/academy/springbootwebgreqit/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+                .anyRequest().authenticated()
+            )
+            .formLogin(form -> form
+                .loginPage("/login")
+                .permitAll()
+            )
+            .logout(logout -> logout
+                .permitAll()
+            );
+        return http.build();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/config/SecurityConfig.java:33]:
```suggestion
            .sessionManagement(
                        s -> s.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
                .userDetailsService(service)
                .build();
```

File: src/main/java/mate/academy/springbootwebgreqit/controller/AdminController.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/admin")
+public class AdminController {
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @GetMapping("/dashboard")
+    public String getAdminDashboard() {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/controller/AdminController.java:13]: did you see the tutorials from security module?

Pleasem check task description again and provide implementation according to it

File: src/main/java/mate/academy/springbootwebgreqit/controller/AdminController.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package mate.academy.springbootwebgreqit.controller;
+
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/admin")
+public class AdminController {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/controller/AdminController.java:10]:
```suggestion
@RequestMapping("/auth")
public class AuthenticationController {
```

File: src/main/java/mate/academy/springbootwebgreqit/model/Book.java [hunk 6]
```diff
@@ -31,5 +31,6 @@ public class Book {
     private String description;
     @Column(nullable = false, length = 255)
     private String coverImage;
+    @Column(name = "is_deleted")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/Book.java:34]:
```suggestion
```

File: src/main/java/mate/academy/springbootwebgreqit/model/Role.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package mate.academy.springbootwebgreqit.model;
+
+import jakarta.persistence.*;
+import lombok.Data;
+import mate.academy.springbootwebgreqit.roles.RoleName;
+import org.springframework.security.core.GrantedAuthority;
+
+@Data
+@Entity
+public abstract class Role implements GrantedAuthority {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/Role.java:10]:
```suggestion
public class Role implements GrantedAuthority {
```

File: src/main/java/mate/academy/springbootwebgreqit/model/Role.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package mate.academy.springbootwebgreqit.model;
+
+import jakarta.persistence.*;
+import lombok.Data;
+import mate.academy.springbootwebgreqit.roles.RoleName;
+import org.springframework.security.core.GrantedAuthority;
+
+@Data
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/Role.java:8]:
```suggestion
```

Add Table, get and set annotations

File: src/main/java/mate/academy/springbootwebgreqit/roles/RoleName.java [hunk 9]
```diff
@@ -0,0 +1,6 @@
+package mate.academy.springbootwebgreqit.roles;
+
+public enum RoleName {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/roles/RoleName.java:3]: you can create this enum in Role class

File: src/main/java/mate/academy/springbootwebgreqit/model/Role.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+@Entity
+public abstract class Role implements GrantedAuthority {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Enumerated(EnumType.STRING)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/Role.java:15]: Add required constraint

File: src/main/java/mate/academy/springbootwebgreqit/model/User.java [hunk 11]
```diff
@@ -2,14 +2,22 @@
...
+import org.springframework.security.core.userdetails.UserDetails;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
 
 @Getter
 @Setter
 @Entity
+@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id = ?")
+@Where(clause = "is_deleted = false")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/User.java:18]: `@Where` is deprecated

File: src/main/java/mate/academy/springbootwebgreqit/model/User.java [hunk 12]
```diff
@@ -18,8 +26,53 @@ public class User {
...
+    private String firstName;
     @Column(nullable = false)
-    private String LastName;
+    private String lastName;
     private String shippingAddress;
-}
+    @Column(name = "is_deleted")
+    private boolean isDeleted = false;
+
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/User.java:36]:
```suggestion
    @ManyToMany
```

File: src/main/java/mate/academy/springbootwebgreqit/model/User.java [hunk 13]
```diff
@@ -18,8 +26,53 @@ public class User {
...
+    @JoinTable(
+            name = "user_roles",
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/User.java:38]:
```suggestion
            name = "users_roles",
```

File: src/main/java/mate/academy/springbootwebgreqit/model/User.java [hunk 14]
```diff
@@ -18,8 +26,53 @@ public class User {
...
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/User.java:42]:
```suggestion
    private Set<Role> roles = new HashSet<>();
```

File: src/main/java/mate/academy/springbootwebgreqit/model/User.java [hunk 15]
```diff
@@ -18,8 +26,53 @@ public class User {
...
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return List.of();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/User.java:46]:
```suggestion
        return roles;
```

File: src/main/java/mate/academy/springbootwebgreqit/model/User.java [hunk 16]
```diff
@@ -18,8 +26,53 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return !isDeleted;
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return UserDetails.super.isEnabled();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/User.java:71]:
```suggestion
        return !isDeleted;
```

File: src/main/java/mate/academy/springbootwebgreqit/model/User.java [hunk 17]
```diff
@@ -18,8 +26,53 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isAccountNonLocked() {
+        return true;
+    }
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return !isDeleted;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/User.java:66]:
```suggestion
        return true;
```

File: src/main/java/mate/academy/springbootwebgreqit/repository/RoleRepository.java [hunk 18]
```diff
@@ -0,0 +1,7 @@
+package mate.academy.springbootwebgreqit.repository;
+
+import mate.academy.springbootwebgreqit.model.Role;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public abstract class RoleRepository implements JpaRepository<Role, Long> {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/repository/RoleRepository.java:6]: abstract class repository?

File: src/main/java/mate/academy/springbootwebgreqit/security/CustomUserDetailService.java [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailService implements UserDetailsService {
+    private  final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        User user = userRepository.findByEmail(email)
+                .orElseThrow(() -> new UsernameNotFoundException("Can't find user by email"));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/security/CustomUserDetailService.java:19]: use custom entity not found exception

File: src/main/java/mate/academy/springbootwebgreqit/security/CustomUserDetailService.java [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailService implements UserDetailsService {
+    private  final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        User user = userRepository.findByEmail(email)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/security/CustomUserDetailService.java:18]:
```suggestion
        return userRepository.findByEmail(email)
```

File: src/main/resources/db/changelog/changes/02-add-change-role.yaml [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+                  constraints:
+                    nullable: false
+                    unique: true
+
+  - changeSet:
+      id: 2
+      author: qreqit
+      changes:
+        - createTable:
+            tableName: user_roles
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-add-change-role.yaml:28]:
```suggestion
            tableName: users_roles
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: pom.xml [hunk 22]
```diff
@@ -50,6 +51,13 @@
             <version>3.3.0</version>
         </dependency>
 
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-security</artifactId>
+            <version>${spring-boot-starter-security.version}</version>
```
Comment (Elena-Bruyako) [inline pom.xml:57]:
```suggestion
```

File: src/main/java/mate/academy/springbootwebgreqit/config/SecurityConfig.java [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.security.config.Customizer;
+import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@EnableMethodSecurity
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/config/SecurityConfig.java:14]:
```suggestion
@EnableMethodSecurity
@RequiredArgsConstructor
```

File: src/main/java/mate/academy/springbootwebgreqit/config/SecurityConfig.java [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+@Configuration
+public class SecurityConfig {
+    private UserDetailsService userDetailsService;
+
+    public SecurityConfig(UserDetailsService userDetailsService) {
+        this.userDetailsService = userDetailsService;
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/config/SecurityConfig.java:21]:
```suggestion
    private final UserDetailsService userDetailsService;
```

File: src/main/java/mate/academy/springbootwebgreqit/controller/AuthenticationController.java [hunk 25]
```diff
@@ -1,22 +1,16 @@
...
-@RequiredArgsConstructor
 @RestController
 @RequestMapping("/auth")
 public class AuthenticationController {
-    private final UserService userService;
-
-    @PostMapping("/registration")
-    public UserResponseDto registerUser(@RequestBody UserRegistrationRequestDto requestBody) {
-        return userService.register(requestBody);
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/controller/AuthenticationController.java:11]: Why did you make this changes?

File: src/main/java/mate/academy/springbootwebgreqit/model/Role.java [hunk 26]
```diff
@@ -0,0 +1,25 @@
+package mate.academy.springbootwebgreqit.model;
+
+import jakarta.persistence.*;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/Role.java:6]: Add Table annotation

File: src/main/java/mate/academy/springbootwebgreqit/model/User.java [hunk 27]
```diff
@@ -18,8 +29,56 @@ public class User {
...
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles = new HashSet<>();
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        if (email.startsWith("admin@")) {
+            return List.of(new SimpleGrantedAuthority("ROLE_ADMIN"));
+        }
+        return List.of(new SimpleGrantedAuthority("ROLE_USER"));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/User.java:52]:
```suggestion
        return roles;
```

File: src/main/resources/db/changelog/changes/02-add-change-role.yaml (hunk 21 above)
Comment (Elena-Bruyako) [reply src/main/resources/db/changelog/changes/02-add-change-role.yaml:28]: Not fixed

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/mate/academy/springbootwebgreqit/controller/AuthenticationController.java (hunk 25 above)
Comment (qreqit) [reply src/main/java/mate/academy/springbootwebgreqit/controller/AuthenticationController.java:11]: ohh, I made a mistake

File: src/main/java/mate/academy/springbootwebgreqit/model/Role.java [hunk 28]
```diff
@@ -0,0 +1,26 @@
+package mate.academy.springbootwebgreqit.model;
+
+import jakarta.persistence.*;
+import org.springframework.security.core.GrantedAuthority;
+
+@Table
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootwebgreqit/model/Role.java:6]: name ?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
one comment

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**